### PR TITLE
fix(#7250): race condition in loading contacts

### PR DESCRIPTION
**Pull request by latin-panda**
# Description

Stores the contact ID we want to fetch. Then as promises resolve, it checks that the contact we want to load is the same as from the data we had retrieved. 

medic/cht-core#7250

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:7250-Race-condition-in-loading-contacts/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:7250-Race-condition-in-loading-contacts/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:7250-Race-condition-in-loading-contacts/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by latin-panda** [comment]
@ngaruko and @dianabarsan 

There's a [job before on this topic](https://github.com/medic/cht-core/commit/bf36dd4cb543ee7d183dc2f7dcb4287144c880ba). The only part where we still need to check for the ID that it's loading is after the [first retrieve of data](https://github.com/medic/cht-core/blob/45eb3e23abe57c945656f4f6226054d583c5e4bd/webapp/src/ts/effects/contacts.effects.ts#L97).

Please have a look and let me know your thoughts.

File: webapp/src/ts/actions/contacts.ts [hunk 1]
```diff
@@ -9,6 +9,7 @@ export const Actions = {
   selectContact: createMultiValueAction('SELECT_CONTACT'),
   setSelectedContact: createSingleValueAction('SET_SELECTED_CONTACT', 'selected'),
   setContactsLoadingSummary: createSingleValueAction('SET_CONTACT_LOADING_SUMMARY', 'value'),
+  setContactIdToFetch: createSingleValueAction('SET_CONTACT_ID_TO_FETCH', 'id'),
```
Comment (dianabarsan) [inline webapp/src/ts/actions/contacts.ts:12]: For consistency (we don't use the term "fetch" anywhere else), could you please rename this to something like `setContactIdToSelect` ?

File: webapp/src/ts/effects/contacts.effects.ts [hunk 2]
```diff
@@ -82,26 +95,34 @@ export class ContactsEffects {
...
 
-  private setTitle(selected) {
-    const routeSnapshot = this.routeSnapshotService.get();
-    const deceasedTitle = routeSnapshot?.data?.name === 'contacts.deceased'
-      ? this.translateService.instant('contact.deceased.title') : null;
-    const title = deceasedTitle || selected.type?.name_key || 'contact.profile';
-    this.globalActions.setTitle(this.translateService.instant(title));
+  private setTitle(contactId) {
+    return this
+      .verifySelectedContactNotChanged(contactId)
```
Comment (dianabarsan) [inline webapp/src/ts/effects/contacts.effects.ts:100]: I wonder if we can move this to the check to the main promise chain (to line 67) and keep this function cleaner and doing what it's called (sets the title).

**Review by dianabarsan — CHANGES_REQUESTED** [review]
oops I accidentally hit enter without filling the textarea! 

Thanks a lot for this change, @latin-panda ! It's great! I added a small request to keep code cleaner and a renaming suggestion.

File: webapp/src/ts/actions/contacts.ts (hunk 1 above)
Comment (latin-panda) [reply webapp/src/ts/actions/contacts.ts:12]: Yeah, the name is weird. I want to make a difference when it looks for contact data and when it sets the contact as "selected" once we have the data so it's more straightforward for devs when it's used. 

Maybe `setContactIdToFind`? or `setContactIdToSummon`? 😅

Comment (dianabarsan) [reply webapp/src/ts/actions/contacts.ts:12]: `setContactIdToLoad` ?

File: webapp/src/ts/effects/contacts.effects.ts (hunk 2 above)
Comment (latin-panda) [reply webapp/src/ts/effects/contacts.effects.ts:100]: Done

File: webapp/src/ts/actions/contacts.ts (hunk 1 above)
Comment (latin-panda) [reply webapp/src/ts/actions/contacts.ts:12]: Done

**Comment by latin-panda** [comment]
@ngaruko I found the steps to reproduce the issue reliably; it's in [the ticket](https://github.com/medic/cht-core/issues/7250#issuecomment-1543777876).

**Review by dianabarsan — APPROVED** [review]
Nice, these things are such a pain. i think we should definitely normalize this kind of flow in all other areas where you could get subsequent select events, before initial ones finish executing.

**Pull request merged**